### PR TITLE
refactor(compiler): parse bindings "by hand" rather than via regex

### PR DESCRIPTION
**Pull request by ayazhafiz**
To support recovery of malformed binding property names like `([a)`,
`[a`, or `()`, the binding parser needs to be more permissive w.r.t. the
kinds of bindings it can detect. This is difficult to do maintainably
with a regex, but is trivial with a "hand-rolled" string parser. This
commit refactors render3's binding attribute parsing to use this method
for multi-delimited bindings (namely via the `()`, `[]`, and `[()]`)
syntax, making the way recovery of malformed bindings in a future patch.

Note that we can keep using a regex for prefix-only binding syntax
(e.g. `bind-`, `ref-`) because validation of the binding is complete
once we have matched the prefix, and the only thing left to do is check
that the binding identifier is non-empty, which is trivial.

Part of #38596

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 1]
```diff
@@ -380,36 +370,58 @@ class HtmlAstToIvyAst implements html.Visitor {
...
+    } else if (name.startsWith(DELIMS.EVENT.start)) {
+      delims = DELIMS.EVENT;
+    }
+    if (delims &&
+        // NB: older versions of the parser would match a start/end delimited
+        // binding iff the property name was terminated by the ending delimiter
+        // and the identifier in the binding was non-empty.
+        // TODO(ayazhafiz): update this to handle malformed bindings.
+        name.endsWith(delims.end) && name.length > delims.start.length + delims.end.length) {
+      const ident = name.substring(delims.start.length, name.length - delims.end.length);
```
Comment (atscott) [inline packages/compiler/src/render3/r3_template_transform.ts:398]: nit: unless the short name is super common/obvious, use a full name for the variable (I'm guessing `identifier` here)

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 2]
```diff
@@ -380,36 +370,58 @@ class HtmlAstToIvyAst implements html.Visitor {
...
+      delims = DELIMS.BANANA_BOX;
+    } else if (name.startsWith(DELIMS.PROPERTY.start)) {
+      delims = DELIMS.PROPERTY;
+    } else if (name.startsWith(DELIMS.EVENT.start)) {
+      delims = DELIMS.EVENT;
+    }
+    if (delims &&
+        // NB: older versions of the parser would match a start/end delimited
+        // binding iff the property name was terminated by the ending delimiter
+        // and the identifier in the binding was non-empty.
```
Comment (atscott) [inline packages/compiler/src/render3/r3_template_transform.ts:395]:
```suggestion
        // Note: older versions of the parser would match a start/end delimited
        // binding iff the property name was terminated by the ending delimiter
        // and the identifier in the binding was non-empty.
```
I admit to having to look up what "NB" stood for. Can we just use "Note:" instead?

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 3]
```diff
@@ -380,36 +370,58 @@ class HtmlAstToIvyAst implements html.Visitor {
...
+        const keySpan = createKeySpan(srcSpan, delims.start, ident);
         this.bindingParser.parsePropertyBinding(
-            bindParts[IDENT_PROPERTY_IDX], value, false, srcSpan, absoluteOffset,
-            attribute.valueSpan, matchableAttributes, parsedProperties, keySpan);
-
-      } else if (bindParts[IDENT_EVENT_IDX]) {
+            ident, value, false, srcSpan, absoluteOffset, attribute.valueSpan, matchableAttributes,
+            parsedProperties, keySpan);
+      } else {
         const events: ParsedEvent[] = [];
```
Comment (atscott) [inline packages/compiler/src/render3/r3_template_transform.ts:412]: Oh boy, I do not like events gets passed in and mutated. But I digress...

**Review by atscott — APPROVED** [review]
👍

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 4]
```diff
@@ -15,13 +15,11 @@ import {ParseError, ParseErrorLevel, ParseSourceSpan} from '../parse_util';
...
 import {BindingParser} from '../template_parser/binding_parser';
 import {PreparsedElementType, preparseElement} from '../template_parser/template_preparser';
-import {syntaxError} from '../util';
 
 import * as t from './r3_ast';
 import {I18N_ICU_VAR_PREFIX, isI18nRootNode} from './view/i18n/util';
 
-const BIND_NAME_REGEXP =
-    /^(?:(?:(?:(bind-)|(let-)|(ref-|#)|(on-)|(bindon-)|(@))(.*))|\[\(([^\)]+)\)\]|\[([^\]]+)\]|\(([^\)]+)\))$/;
+const BIND_NAME_REGEXP = /^(?:(?:(bind-)|(let-)|(ref-|#)|(on-)|(bindon-)|(@))(.*))$/;
```
Comment (gkalpak) [inline packages/compiler/src/render3/r3_template_transform.ts:22]: Drive-by comment (not directly related to this PR either :grin:):
The outer non-capturing parenthesis is redundant:

```suggestion
const BIND_NAME_REGEXP = /^(?:(bind-)|(let-)|(ref-|#)|(on-)|(bindon-)|(@))(.*)$/;
```

File: packages/compiler/src/render3/r3_template_transform.ts (hunk 1 above)
Comment (ayazhafiz) [reply packages/compiler/src/render3/r3_template_transform.ts:398]: I used this because it was used elsewhere in the parser, and I have updated it :)

File: packages/compiler/src/render3/r3_template_transform.ts (hunk 2 above)
Comment (ayazhafiz) [reply packages/compiler/src/render3/r3_template_transform.ts:395]: Yep, done

File: packages/compiler/src/render3/r3_template_transform.ts (hunk 4 above)
Comment (ayazhafiz) [reply packages/compiler/src/render3/r3_template_transform.ts:22]: Nice, done!

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 5]
```diff
@@ -380,36 +370,59 @@ class HtmlAstToIvyAst implements html.Visitor {
...
             parsedProperties, keySpan);
+      }
+      return true;
+    }
 
-      } else if (bindParts[IDENT_BANANA_BOX_IDX]) {
-        const keySpan = createKeySpan(srcSpan, '[(', bindParts[IDENT_BANANA_BOX_IDX]);
+    // We didn't see a kw-prefixed property binding, but we have not yet checked
+    // for the []/()/[()] syntax.
+    const DELIMS = {
```
Comment (alxhub) [inline packages/compiler/src/render3/r3_template_transform.ts:379]: This could be extracted to a top-level constant, to avoid allocating it on every run.

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 6]
```diff
@@ -380,36 +370,59 @@ class HtmlAstToIvyAst implements html.Visitor {
...
+      BANANA_BOX: {start: '[(', end: ')]'},
+      PROPERTY: {start: '[', end: ']'},
+      EVENT: {start: '(', end: ')'},
+    };
+    let delims: {start: string, end: string}|undefined;
```
Comment (alxhub) [inline packages/compiler/src/render3/r3_template_transform.ts:384]: Prefer `null` and explicit initialization/comparison

**Review by alxhub — APPROVED** [review]

**Review by alxhub — APPROVED** [review]

**Comment by ayazhafiz** [comment]
Caretaker: can you run g3sync?
